wgengine/magicsock/magicsock_test.go

```diff
@@ -3098,14 +3092,56 @@ func TestBlockEndpointsDERPOK(t *testing.T) {
 		break
 	}
 
-	cleanup = newPinger(t, t.Logf, ms1, ms2)
-	defer cleanup()
+	waitForEndpoints(t, ms1.conn)
+	waitForEndpoints(t, ms2.conn)
 
-	if len(ms1.conn.activeDerp) == 0 {
-		t.Errorf("unexpected DERP empty got: %v want: >0", len(ms1.conn.activeDerp))
+	// meshStacks doesn't use call-me-maybe packets to update endpoints, which
+	// makes them not get cleared when we call SetBlockEndpoints.
+	ep2, ok := ms1.conn.peerMap.endpointForNodeKey(ms2.Public())
+	if !ok {
+		t.Fatalf("endpoint not found for ms2")
+	}
+	ep2.mu.Lock()
+	for k := range ep2.endpointState {
+		ep2.deleteEndpointLocked("TestBlockEndpointsDERPOK", k)
+	}
+	ep2.mu.Unlock()
+
+	// Wait for the call-me-maybe packet to arrive.
+	for i := 0; i < 50; i++ {
+		time.Sleep(100 * time.Millisecond)
+		ep2.mu.Lock()
+		if len(ep2.endpointState) != 0 {
+			ep2.mu.Unlock()
+			break
+		}
+		ep2.mu.Unlock()
 	}
-	if len(ms2.conn.activeDerp) == 0 {
-		t.Errorf("unexpected DERP empty got: %v want: >0", len(ms2.conn.activeDerp))
+
+	// SetBlockEndpoints is called later since it's incompatible with the test
+	// meshStacks implementations.
+	// We only set it on ms1, since ms2's endpoints should be ignored by ms1.
+	ms1.conn.SetBlockEndpoints(true)
+
+	// All call-me-maybe endpoints should've been immediately removed from ms1.
+	ep2.mu.Lock()
+	if len(ep2.endpointState) != 0 {
+		ep2.mu.Unlock()
+		t.Fatalf("endpoints not removed from ms1")
+	}
+	ep2.mu.Unlock()
+
+	// Wait for endpoints to finish updating.
+	waitForNoEndpoints(t, ms1.conn)
+
+	// Give time for another call-me-maybe packet to arrive. I couldn't think of
+	// a better way than sleeping without making a bunch of changes.
+	time.Sleep(time.Second)
```

Comment:
I had to set `ms1.conn.SetDebugLoggingEnabled(true)` and add logs before and after `time.Sleep()` to convince myself that call-me-maybe packets do get sent during this sleep.  I suggest we add this to the test.

I don't have a better idea for testing that's not a lot of work.